### ISSUE TITLE
Train DNN model efficiently with sparse features and missing data

### ISSUE THREAD

**Issue by tobegit3hub** [tensorflow/tensorflow#4114]
We have used TensorFlow for recommend systems to replace the logistic regression model. The train data is quite sparse and we have to do it efficiently. Now we have the problem to train the model without sparse tensor's multiplication.

We have read `word2vec` and the example of `wide_n_deep` models. Their inputs are sparse but not suitable for general recommend systems. Our inputs are sparse and we need to encode the features like ages, colleges with one-hot encoding. The examples may have different number of valid features because of unknown values and the train data looks like this.

```
label         gender        age        college       
------------------------------------------------
0             2:1           8:1       (unknown)
1             1:1           5:1         25:1
0             1:1           8:1       (unknown)
```

With this dataset, we have to build `SparseTensor` object for each example data like this.

```
example1 = tf.SparseTensor(indices=[[2], [8]], values=[1, 1], shape=[100])
example2 = tf.SparseTensor(indices=[[1], [5], [25]], values=[1, 1, 1], shape=[100])
example3 = tf.SparseTensor(indices=[[1], [8]], values=[1, 1], shape=[100])
```

For `word2vec` and `wide_n_deep` models, we can use `tf.nn.embedding_lookup()` to lookup the variables to train and no need to fill up the whole matrix with zeros. This works for each example but we have problems if using batch because the valid `ids` has different shape to train. The code  looks like this. 

```
vocabulary_size = 100
embedding_size = 1
embeddings = tf.Variable(tf.ones([vocabulary_size, embedding_size]))

batch_size = 3
feature_number = 3 # ERROR: should be 2 or 3
train_inputs = tf.placeholder(tf.int32, shape=[batch_size, feature_number])

batch_data = np.array([[2, 8], [1, 5, 25], [1, 8]]) # EROOR: should be dense

with tf.Session() as sess:
    sess.run(tf.initialize_all_variables())
    print(sess.run(embed, feed_dict={train_inputs: batch_data}))
```

The code could not work until we make `example1`, `example2` and `example3` have the same number of valid featue, such as 3 in this case. This is one solution but we have to fill up the examples with zeros(notice that this is different from filling up the one-hot encoding).

Maybe supporting sparse tensor's multiplication is quit more efficient. Now we have only `embedding_lookup` op for SpareTensor and I don't know how to use `embedding_lookup_sparse` or `safe_embedding_lookup_sparse` for this scenario. It could be great if anyone has any suggestion for this use case.

Related to https://github.com/tensorflow/tensorflow/issues/1241

**Comment by tobegit3hub** [comment]
@ebrevdo @mrry can you guys help?

I thinks it's an general scenario but we have no idea to implement it with TensorFlow. We have implemented the logistic regression model with the same dataset in Spark. It iterates each example without batching and doesn't have this issue. We can also fill up the `unknown` data with zeros or default values but this is some kind of inefficient because actually we don't need to compute their variables.

**Comment by ebrevdo** [comment]
You can create a rank-3 SparseTensor that contains all batches.  Its indices matrix would look like, e.g. (from your example):

```
[0  2]
[0  8]
[1  1]
[1  5]
[1 25]
[2  1]
[2  8]
```
and its values would be a vector with the concatenated values of the 3 batch entries.

`safe_embedding_lookup_sparse` can handle this kind of SparseTensor (actually i think the others can too?)

Does this answer your question?  I'm not 100% sure.

**Comment by ebrevdo** [comment]
(there's also the `sparse_tensor_dense_matmul` op which allows one to multiply rank-2 SparseTensors with dense matrices; but i feel that `safe_embedding_lookup_sparse` is sufficient for what you want to do)

**Comment by tobegit3hub** [comment]
Thanks @ebrevdo for the quick response. We have considered to use the rank-3 SparseTensor instead of a batch of rank-2 SparseTensors. It's the different way to represent the train data and of course we can use it.

But the `safe_embedding_lookup_sparse` is not only the lookup method but also combining the weights sum/average of each row. I have read the [api docs](https://www.tensorflow.org/versions/r0.10/api_docs/python/nn.html#embedding_lookup_sparse) and trace the unit test of `safe_embedding_lookup_sparse`.

![screen shot 2016-08-30 at 12 19 02](https://cloud.githubusercontent.com/assets/2715000/18116453/89381de0-6f7a-11e6-9635-26ef332160d3.png)

It's some kind of computation but not for our model. For dense data, we compute the logist with `w * x + b` where w, x and b are dense matrix. For sparse data, just like word2vev, we use `embedding_lookup` to lookup the related variables and apply to the similar function. Now we can do that but `embedding_lookup` need the dense `ids` and we can't satisfy this because we have missing features now.

We have tried `safe_embedding_lookup_sparse` but don't know any way to do something like `w * x + b`. It would be better to have an example about using this to get the logis so that we can write the loss function to train.

Not sure if `safe_embedding_lookup_sparse` can do this. If not, we need to lookup each "reldated"  variables and do the sparse multiplication.

**Comment by tobegit3hub** [comment]
[Wide and deep model](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/examples/learn/wide_n_deep_tutorial.py) is the best example to handle sparse data and use `safe_embedding_lookup_sparse`. Notice that the input data doesn't have missing feature, which means each example has exactly same number of feature values.

The model handles continues columns and categorical(sparse) columns in different ways. It will create one SparseTensor for each categorical column, which is quite different from what @ebrevdo  and I mentioned above. Now we have many SparseTensors such as "education_num", "capital_gain" and "capital_loss" and the shape is the same `[batch_size, 1]`. This is the precondition of using `safe_embedding_lookup_sparse` because it doesn't need to aggregate anything in this use case. In each train step, we need a `for` loop to iterate add all columns' outputs and it will be much slow if we have much more features.

So I don't know `safe_embedding_lookup_sparse` is the solution for this problem. It works like `embedding_lookup` and we have to create many SparseTensors for each column.

It would be better to know how Google uses TensorFlow for sparse inputs internally. It's general in recommendation system and CTR scenario. Because what I want is how to handle this use cause instead of using sparse matrix multiplication, I'm gonna change the tittle and please comment if you have any practices with TensorFlow.

**Comment by aselle** [comment]
This question, since it concerns how to use TensorFlow properly, is much better asked on StackOverflow. If you believe there is a bug or a very specific feature request, please resubmit those as more focused requests referencing this issue.

**Comment by automateljw** [comment]
@tobegit3hub  do you have resolved the problem?

**Comment by 9162726598** [comment]
could tensorflow 2.0 support sparse input training and prediction now?
